File: templates/Uwp/Pages/WebViewVB/ViewModels/WebViewPageViewModel.vb
Namespace ViewModels
    Public Class WebViewPageViewModel
        Inherits System.ComponentModel.INotifyPropertyChanged

        ' TODO WTS: Set the URI of the page to show by default
        Private Const DefaultUrl As String = "https://developer.microsoft.com/en-us/windows/apps"

        Private _source As Uri

        Public Property Source As Uri
            Get
                Return _source
            End Get
            Set
                [Param_Setter](_source, value)
            End Set
        End Property

        Private _isLoading As Boolean

        Public Property IsLoading As Boolean
            Get
                Return _isLoading
            End Get

            Set
                If value Then
                    IsShowingFailedMessage = False
                End If

                [Param_Setter](_isLoading, value)
                IsLoadingVisibility = If(value, Visibility.Visible, Visibility.Collapsed)
            End Set
        End Property

        Private _isLoadingVisibility As Visibility

        Public Property IsLoadingVisibility As Visibility
            Get
                Return _isLoadingVisibility
            End Get
            Set
                [Param_Setter](_isLoadingVisibility, value)
            End Set
        End Property

        Private _isShowingFailedMessage As Boolean

        Public Property IsShowingFailedMessage As Boolean
            Get
                Return _isShowingFailedMessage
            End Get

            Set
                If value Then
                    IsLoading = False
                End If

                [Param_Setter](_isShowingFailedMessage, value)
                FailedMesageVisibility = If(value, Visibility.Visible, Visibility.Collapsed)
            End Set
        End Property

        Private _failedMesageVisibility As Visibility

        Public Property FailedMesageVisibility As Visibility
            Get
                Return _failedMesageVisibility
            End Get
            Set
                [Param_Setter](_failedMesageVisibility, value)
            End Set
        End Property

        Private _navCompleted As ICommand

        Public ReadOnly Property NavCompletedCommand As ICommand
            Get
                If _navCompleted Is Nothing Then
                    _navCompleted = New RelayCommand(Of WebViewNavigationCompletedEventArgs)(AddressOf NavCompleted)
                End If

                Return _navCompleted
            End Get
        End Property

        Private Sub NavCompleted(e As WebViewNavigationCompletedEventArgs)
            IsLoading = False
        End Sub

        Private _navFailed As ICommand

        Public ReadOnly Property NavFailedCommand As ICommand
            Get
                If _navFailed Is Nothing Then
                    _navFailed = New RelayCommand(Of WebViewNavigationFailedEventArgs)(AddressOf NavFailed)
                End If

                Return _navFailed
            End Get
        End Property

        Private Sub NavFailed(e As WebViewNavigationFailedEventArgs)
            ' Use `e.WebErrorStatus` to vary the displayed message based on the error reason
            IsShowingFailedMessage = True
        End Sub

        Private _retryCommand As ICommand

        Public ReadOnly Property RetryCommand As ICommand
            Get
                If _retryCommand Is Nothing Then
                    _retryCommand = New RelayCommand(AddressOf Retry)
                End If

                Return _retryCommand
            End Get
        End Property

        Private Sub Retry()
            IsShowingFailedMessage = False
            IsLoading = True

            _webView?.Refresh()
        End Sub

        Private _browserBackCommand As ICommand

        Public ReadOnly Property BrowserBackCommand As ICommand
            Get
                If _browserBackCommand Is Nothing Then
                    _browserBackCommand = New RelayCommand(Sub() _webView?.GoBack(), Function() If(_webView?.CanGoBack, False))
                End If

                Return _browserBackCommand
            End Get
        End Property

        Private _browserForwardCommand As ICommand

        Public ReadOnly Property BrowserForwardCommand As ICommand
            Get
                If _browserForwardCommand Is Nothing Then
                    _browserForwardCommand = New RelayCommand(Sub() _webView?.GoForward(), Function() If(_webView?.CanGoForward, False))
                End If

                Return _browserForwardCommand
            End Get
        End Property

        Private _refreshCommand As ICommand

        Public ReadOnly Property RefreshCommand As ICommand
            Get
                If _refreshCommand Is Nothing Then
                    _refreshCommand = New RelayCommand(Sub() _webView?.Refresh())
                End If

                Return _refreshCommand
            End Get
        End Property

        Private _openInBrowserCommand As ICommand

        Public ReadOnly Property OpenInBrowserCommand As ICommand
            Get
                If _openInBrowserCommand Is Nothing Then
                    _openInBrowserCommand = New RelayCommand(Async Sub() Await Windows.System.Launcher.LaunchUriAsync(Source))
                End If

                Return _openInBrowserCommand
            End Get
        End Property

        Private _webView As WebView

        Public Sub New()
            IsLoading = True
            Source = New Uri(DefaultUrl)
        End Sub

        Public Sub Initialize(webView As WebView)
            _webView = webView
        End Sub
    End Class
End Namespace

File: code/TemplateStudioForUWP/Templates/Pg/WebView.Prism/ViewModels/WebViewPageViewModel.cs
using System;
using System.Windows.Input;
using Prism.Commands;
using Prism.Windows.Mvvm;
using Windows.UI.Xaml;
using Windows.UI.Xaml.Controls;
using Param_RootNamespace.Services;

namespace Param_RootNamespace.ViewModels
{
    public class WebViewPageViewModel : ViewModelBase
    {
        // TODO WTS: Set the URI of the page to show by default
        private const string DefaultUrl = "https://docs.microsoft.com/windows/apps/";

        public WebViewPageViewModel()
        {
            IsLoading = true;
            Source = new Uri(DefaultUrl);

            BrowserBackCommand = new DelegateCommand(() => _webViewService?.GoBack(), () => _webViewService?.CanGoBack ?? false);
            BrowserForwardCommand = new DelegateCommand(() => _webViewService?.GoForward(), () => _webViewService?.CanGoForward ?? false);
            RefreshCommand = new DelegateCommand(() => _webViewService?.Refresh());
            RetryCommand = new DelegateCommand(Retry);
            OpenInBrowserCommand = new DelegateCommand(async () => await Windows.System.Launcher.LaunchUriAsync(Source));

            // Note that the WebViewService is set from within the view because it needs a reference to the WebView control
        }

        private Uri _source;

        public Uri Source
        {
            get { return _source; }
            set { Param_Setter(ref _source, value); }
        }

        private bool _isLoading;

        public bool IsLoading
        {
            get
            {
                 return _isLoading;
            }

            set
            {
                if (value)
                {
                    IsShowingFailedMessage = false;
                }

                Param_Setter(ref _isLoading, value);
                IsLoadingVisibility = value ? Visibility.Visible : Visibility.Collapsed;
            }
        }

        private Visibility _isLoadingVisibility;

        public Visibility IsLoadingVisibility
        {
            get
            {
                return _isLoadingVisibility;
            }

            set
            {
                Param_Setter(ref _isLoadingVisibility, value);
            }
        }

        private bool _isShowingFailedMessage;

        public bool IsShowingFailedMessage
        {
            get
            {
                return _isShowingFailedMessage;
            }

            set
            {
                if (value)
                {
                    IsLoading = false;
                }

                Param_Setter(ref _isShowingFailedMessage, value);
                FailedMesageVisibility = value ? Visibility.Visible : Visibility.Collapsed;
            }
        }

        private Visibility _failedMessageVisibility;

        public Visibility FailedMesageVisibility
        {
            get
            {
                return _failedMessageVisibility;
            }

            set
            {
                Param_Setter(ref _failedMessageVisibility, value);
            }
        }

        private IWebViewService _webViewService;

        public IWebViewService WebViewService
        {
            get
            {
                return _webViewService;
            }

            // the WebViewService is set from within the view (instead of IoC) because it needs a reference to the control
            set
            {
                if (_webViewService != null)
                {
                    _webViewService.NavigationComplete -= WebViewService_NavigationComplete;
                    _webViewService.NavigationFailed -= WebViewService_NavigationFailed;
                }

                _webViewService = value;
                _webViewService.NavigationComplete += WebViewService_NavigationComplete;
                _webViewService.NavigationFailed += WebViewService_NavigationFailed;
            }
        }

        public ICommand BrowserBackCommand { get; }

        public ICommand BrowserForwardCommand { get; }

        public ICommand RefreshCommand { get; }

        public ICommand RetryCommand { get; }

        public ICommand OpenInBrowserCommand { get; }

        private void WebViewService_NavigationFailed(object sender, WebViewNavigationFailedEventArgs e)
        {
            NavFailed(e);
        }

        private void WebViewService_NavigationComplete(object sender, WebViewNavigationCompletedEventArgs e)
        {
            NavCompleted(e);
        }

        private void NavCompleted(WebViewNavigationCompletedEventArgs e)
        {
            IsLoading = false;
            RaisePropertyChanged(nameof(BrowserBackCommand));
            RaisePropertyChanged(nameof(BrowserForwardCommand));
        }

        private void NavFailed(WebViewNavigationFailedEventArgs e)
        {
            // Use `e.WebErrorStatus` to vary the displayed message based on the error reason
            IsShowingFailedMessage = true;
        }

        private void Retry()
        {
            IsShowingFailedMessage = false;
            IsLoading = true;

            _webViewService?.Refresh();
        }
    }
}
